src/crypto/CHIPCryptoPAL.h

```diff
@@ -512,6 +513,54 @@ class AesCcm128Key
     uint8_t bytes[kAES_CCM128_Key_Length];
 };
 
+using AesKeyByteArray = uint8_t[CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+
+/**
+ * @brief Platform-specific AES key
+ *
+ * The class represents AES key used by Matter stack (currently, only 128-bit key is supported by
+ * Matter) either in the form of raw key material or key reference, depending on the platform. To
+ * achieve that, it contains an opaque context that can be cast to a specific structure used by
+ * the given platform.
+ */
+class AesKey
+{
+public:
+    AesKey();
+    ~AesKey();
+
+    AesKey(const AesKey &) = delete;
+    AesKey(AesKey &&)      = delete;
+    void operator=(const AesKey &) = delete;
+    void operator=(AesKey &&) = delete;
+
+    CHIP_ERROR Initialize(const AesKeyByteArray & rawKey);
+
+    /**
+     * @brief Get internal context cast to the desired key representation
+     */
+    template <class T>
+    const T & Get() const
+    {
+        return *SafePointerCast<const T *>(&mContext);
+    }
+
+    /**
+     * @brief Get internal context cast to the desired key representation
+     */
+    template <class T>
+    T & Get()
+    {
+        return *SafePointerCast<T *>(&mContext);
+    }
+
+private:
+    struct alignas(size_t) OpaqueContext
+    {
+        uint8_t mOpaque[CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
```

Comment:
What if the opaque version requires more bytes to record metadata for the secure subsystem?

Reply:
Another aspect is that this is currently only applicable to Aes128 keys. If we ever use a different algorithm (e.g. Aes256), there will be lack of API compatibility between library versions, due to the growth of this object which is recast.

BTW, this class basically exactly conflicts with `AesCcm128Key` in the same file. Is there a reason this was not used.

Finally, there is the `SymmetricKeyContext` (still in same file) that was already used to abstract away different methods of encryption for group communication, which also uses AES128 keys, but where we don't want them to be moved around the stack (they reside in GroupDataProvider implementation)

Reply:
This is very similar to other opaque structures in this file. I can add a configuration constant to control the size, which could be adjusted by a platform or whenever we switch to AES-256.

`AesCcm128Key` is used as a raw key material, so it cannot be used for my purpose - using a key as a reference. In fact, I planned to rename `AesCcm128Key` to `Aes[128]KeyMaterial` to make this distinction more clear.

`SymmetricKeyContext`-like approach could be a way to go although:
- it seems like it needlessly introduces another approach to the one used for session encryption in `CryptoContext`. Consequently, we now have two encryption paths without a good reason. IMHO both session and group encryption should only differ in where the keys are retrieved from.
- `SymmetricKeyContext` is not enough - we would like to have a similar key abstractions for HKDF etc (to derive session keys in the secure environment). Of course, we could introduce more interfaces like `SymmetricKeyContext` but for me it seemed cleaner to abstract at the key level rather than the context level because the lower we abstract the more code we reuse across platforms.
- Currently, there seems to be no easy way to inject platform-specific `SymmetricKeyContext` so I guess we would need another factory interface to be able to use either representation of the encryption keys the platform wishes. Do I understand the concept?

I'm open to discussion, but these were my reasons why I decided to abstract at the key level.

Reply:
> Currently, there seems to be no easy way to inject platform-specific SymmetricKeyContext so I guess we would need another factory interface to be able to use either representation of the encryption keys the platform wishes. Do I understand the concept?

Yes. And it's true it then involves management. However, an AesKey *could* implement SymmetricKeyContext.

> it seems like it needlessly introduces another approach to the one used for session encryption in CryptoContext. Consequently, we now have two encryption paths without a good reason. IMHO both session and group encryption should only differ in where the keys are retrieved from.

Yes. I had previously pushed for all messages to use CryptoContext, but there was no time prior to 1.0. Perhaps we now have time.

> AesCcm128Key is used as a raw key material, so it cannot be used for my purpose - using a key as a reference. In fact, I planned to rename AesCcm128Key to Aes[128]KeyMaterial to make this distinction more clear.

It definitely feels like `AesCcm128Key` should disappear